zeebe/dynamic-node-id-provider/src/test/java/io/camunda/zeebe/dynamic/nodeid/repository/s3/S3NodeIdRepositoryIT.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.dynamic.nodeid.repository.s3;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.verify;
+
+import io.camunda.zeebe.dynamic.nodeid.Lease;
+import io.camunda.zeebe.dynamic.nodeid.NodeInstance;
+import io.camunda.zeebe.dynamic.nodeid.repository.NodeIdRepository.StoredLease;
+import io.camunda.zeebe.dynamic.nodeid.repository.s3.S3NodeIdRepository.Config;
+import io.camunda.zeebe.dynamic.nodeid.repository.s3.S3NodeIdRepository.S3ClientConfig;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.Optional;
+import java.util.UUID;
+import org.junit.jupiter.api.AutoClose;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.testcontainers.containers.localstack.LocalStackContainer;
+import org.testcontainers.containers.localstack.LocalStackContainer.Service;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+@Testcontainers
+public class S3NodeIdRepositoryIT {
+
+  private static final Duration EXPIRY_DURATION = Duration.ofSeconds(5);
+
+  @Container
+  private static final LocalStackContainer S3 =
+      new LocalStackContainer(DockerImageName.parse("localstack/localstack:4.10"))
+          .withServices(Service.S3)
+          .withEnv("LS_LOG", "trace");
+
+  @AutoClose private static S3Client client;
+  S3NodeIdRepository.Config config;
+  @AutoClose private S3NodeIdRepository repository;
+
+  @BeforeAll
+  public static void setUpAll() {
+    client =
+        S3NodeIdRepository.buildClient(
+            new S3ClientConfig(
+                S3.getAccessKey(),
+                S3.getSecretKey(),
+                Region.of(S3.getRegion()),
+                Optional.of(S3.getEndpoint())));
+  }
+
+  @BeforeEach
+  public void setUp() {
+    final var bucketName = UUID.randomUUID().toString();
+    final var taskId = UUID.randomUUID().toString();
+    config = new Config(bucketName, taskId, EXPIRY_DURATION);
+    client.createBucket(b -> b.bucket(config.bucketName()));
+  }
+
+  @Test
+  public void shouldInitializeAllFiles() {
```

Comment:
:wrench: Please also add a test that, an acquired lease is not overwritten if we call initialize again.